### PR TITLE
[RFC] Multiple users in an event

### PR DESCRIPTION
**Pull request by webmat**
I'm opening this RFC with a target of stage 2. This is the first PR for this RFC, but a lot of discussion has already happened before on the subject. Notably in #809 and #589 (see this RFC's references links to for more past discussions).

Opening directly at stage 2 simply means we'll need to ensure all stage 1 and 2 criteria are satisfied before moving forward.

In this RFC I try to document all expected uses of the `user` fields. This is not meant to be only a description of the new fields that will be introduced by #869.

Since this RFC is meant to document all uses of the `user` fields, there will also be discussion about potentially deprecating and removing user fields in places that don't have a clear purpose. Feedback is definitely welcome, if you disagree with these removals.

**TODO**

- [x] Identify the sponsor
- [x] Populate the "source data" section
  - [x] add some simple Linux IAM events
  - [x] **I need help** to find more complex IAM event samples (e.g. listing changes to the user)

[Preview of the RFC](https://github.com/webmat/ecs/blob/multiple-users-rfc/rfcs/text/0007-multiple-users.md)

### CODE REVIEW

**Comment by webmat** [comment]
This PR is no longer marked as Draft :-) 

Ping @ebeahan @epixa @jamiehynds @leehinman @janniten @willemdh @neu5ron @vbohata @rw-access @P1llus 

I'll be adding simple Linux events in the "Source Data" section.

However I would really appreciate if someone has real events they could share from AD or other, that capture some of the complex events with many users on the event (e.g. "real" user + priv escalation + target user + changes to the user).

**Comment by jamiehynds** [comment]
@webmat 

**Event ID 4732 in AD - 'administrator' has added 'bob' to the users group.** 

A member was added to a security-enabled local group.
Subject:
   Security ID:  WIN-R9H529RIO4Y\Administrator
   Account Name:  Administrator
   Account Domain:  WIN-R9H529RIO4Y
   Logon ID:  0x1fd23

Member:
   Security ID:  WIN-R9H529RIO4Y\bob
   Account Name:  -

Group:
   Security ID:  BUILTIN\Users
   Group Name:  Users
   Group Domain:  Builtin

Additional Information:

   Privileges:  -

   Expiration time:  -

** Event ID 4722 in AD - 'administrator' has enabled the 'john.locke' account.** 

A user account was enabled.

Subject:

   Security ID:  ACME-FR\administrator
   Account Name:  administrator
   Account Domain:  ACME-FR
   Logon ID:  0x20f9d

Target Account:

   Security ID:  ACME-FR\John.Locke
   Account Name:  John.Locke
   Account Domain:  ACME-FR

**Comment by janniten** [comment]
Hi All,
**AD Example**
I was not able to find an AD event where the 3 users appear. The relationship between users exists but it is scattered in
several events (4648,4624 and 4787) and bound but the winlog.logon.id. For example:
image

I'm working now with some events in which all users maybe appears (4670,4715) in one event. I'll let you know If I found the example that contains all the users.

**Oracle Auditory Event Example**
 Those events are in the table dba_audit_trail in the SYS schema  (here the detailed structure: https://docs.oracle.com/database/121/REFRN/GUID-A9993FAC-12D3-4725-A37D-938CC32D74CC.htm#REFRN23023).
I prepare some small example, copying here the relevant columns but the full export is attached.

OS_USERNAME | USERNAME | ACTION_NAME | SQL_TEXT | Mapping
-- | -- | -- | -- | --
gabriel.surname | SYS | ALTER USER | ALTER USER ACRISTALDI   IDENTIFIED BY * | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI
gabriel.surname | SYS | SYSTEM GRANT | GRANT SELECT ANY   TABLE TO ACRISTALDI | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI
gabriel.surname | SYS | UPDATE | UPDATE SYS.USER$ SET   NAME='ACRISTAL' AND USER#=143 AND NAME='ACRISTALDI'; | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI     **user.changes:** ACRISTAL

[example_oracle.xlsx](https://github.com/elastic/ecs/files/5061746/example_oracle.xlsx)
 
**Privilege Access Management Logs**
In PAM solutions (with the exception of logs regarding to administration of secrets) we always have a case of privilege escalation,  so user.name and user.efective always exists and if the action is related to user management, then user.target and user.changes appears 
For example, I'm coping some headers from Cyberark's logs (unfortunatly I don't have access to export the logs)
image

I have access to some Thycotik logs, but I'm not able to generate logs. I'll be looking for an example there also.
Regards

**Comment by webmat** [comment]
Amazing, thank you @janniten

**Comment by leehinman** [comment]
AWS has an AssumedRole API, a simplified Cloudtrail entry would be:

``` json
{
  "eventName": "AssumeRole",
  "requestParameters": {
    "roleArn": "arn:aws:iam::111111111111:role/JohnRole2",
  },
  "resources": [
    {
      "ARN": "arn:aws:iam::111122223333:role/JohnRole2",
      "accountId": "111111111111",
      "type": "AWS::IAM::Role"
    }
  ],
  "responseElements": {
    "assumedRoleUser": {
      "arn": "arn:aws:sts::111111111111:assumed-role/test-role/Role2WithTags",
      "assumedRoleId": "AROAIFR7WHDTSOYQYHFUE:Role2WithTags"
    },
  "userIdentity": {
    "accessKeyId": "AKIAI44QH8DHBEXAMPLE",
    "accountId": "111111111111",
    "arn": "arn:aws:sts::111111111111:assumed-role/JohnDoe/JohnRole1",
    "principalId": "AROAIN5ATK5U7KEXAMPLE:JohnRole1",
    "type": "AssumedRole"
  }
}
```
| field             | value                               |
|-------------------|-------------------------------------|
| user.id           | AROAIN5ATK5U7KEXAMPLE:JohnRole1     |
| user.effective.id | AROAIFR7WHDTSOYQYHFUE:Role2WithTags |


Also AWS can have AssumedRoles in the userIdentity:

``` json
  "userIdentity": {
    "accessKeyId": "AKIAQRSTUVWXYZEXAMPLE",
    "accountId": "777788889999",
    "arn": "arn:aws:sts::777788889999:assumed-role/AssumeNothing/devdsk",
    "principalId": "AIDAQRSTUVWXYZEXAMPLE:devdsk",
    "sessionContext": {
      "attributes": {
        "creationDate": "2016-11-14T17:25:26Z",
        "mfaAuthenticated": "false"
      },
      "sessionIssuer": {
        "accountId": "777788889999",
        "arn": "arn:aws:iam::777788889999:role/AssumeNothing",
        "principalId": "AIDAQRSTUVWXYZEXAMPLE",
        "type": "Role",
        "userName": "AssumeNothing"
      }
    },
    "type": "AssumedRole"
  }
```
| field             | value                               |
|-------------------|-------------------------------------|
| user.id           | AIDAQRSTUVWXYZEXAMPLE               |
| user.name         | AssumeNothing                       |
| user.effective.id | AIDAQRSTUVWXYZEXAMPLE:devdsk        |

**Comment by jamiehynds** [comment]
Event ID 4624 in Windows is used to capture when a process attempts to log on an account by explicitly specifying that account’s credentials.  This most commonly occurs in batch-type configurations such as scheduled tasks, or when using the RUNAS command.

A logon was attempted using explicit credentials.

Subject:
   Security ID:  WIN-R9H529RIO4Y\Administrator
   Account Name:  **Administrator**
   Account Domain:  WIN-R9H529RIO4Y
   Logon ID:  0x1ba0e
   Logon GUID:  {00000000-0000-0000-0000-000000000000}
Account Whose Credentials Were Used:
   Account Name:  **rsmith@mtg.com**
   Account Domain:  WIN-R9H529RIO4Y
   Logon GUID:  {00000000-0000-0000-0000-000000000000}
Target Server:
   Target Server Name: sp01.IceMAIL.com
   Additional Information: sp01.IceMAIL.com
Process Information:
   Process ID:  0x77c
   Process Name:  C:\Program Files\Internet Explorer\iexplore.exe
Network Information:
   Network Address: -
   Port:   -

File: rfcs/text/0000-multiple-users.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+
+In past discussions and recent research, we have not identified a clear purpose
+for the user fields nested at `host.user.*`.
+
+We are considering deprecating these fields with the intent to remove them completely.
+Please let us know if you disagree with this, and share how you're using them.
+
+### Documenting the purpose of each usage of the user fields
+
+As of ECS 1.5, the ECS documentation doesn't have a good place to explain at length
```
[the remark this answers is not in the document]
Comment (webmat) [reply rfcs/text/0000-multiple-users.md:336]: Makes perfect sense, yes. I'll add that link.

**Comment by jonathan-buttner** [comment]
@webmat just wanted to check in and see if you need me to solicit more feedback (I think I'm the sponsor for this one)?

**Comment by webmat** [comment]
@jonathan-buttner Thanks for asking. Yes, a thorough review from someone on the team would be welcome.

Note that I still have a few tasks on my end before this is final, the main one being including concrete mapping examples based on Linux and Windows events (e.g. like those shared via comments here). This is unlikely to happen this week or the next.

So a review in the meantime would be greatly appreciated, with the only caveat that the concrete examples are coming :-)

File: rfcs/text/0000-multiple-users.md [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+As of ECS 1.6, the ECS documentation doesn't have a good place to explain at length
+how to properly use the multiple nesting locations for `user`.
+This is already a problem for the usage of `user` at the root vs its 5 reuse locations.
+The addition of 3 new reuse locations adds to this situation.
+
+#### Resolution
+
+Adding a way to document field sets via free form text is being worked on independently
+of this proposal ([ecs#943](https://github.com/elastic/ecs/issues/943)).
```
Comment (jonathan-buttner) [inline rfcs/text/0000-multiple-users.md:355]: I wonder if it'd also be useful to have some `description` field or maybe `purpose` within the reusable section to document why the reuse is happening:

```YAML
# schemas/user.yml excerpt
  reusable:
    top_level: true
    expected:
      - client
      - destination
      - host
      - server
      - source
      # Added for this RFC
      - at: user
        as: target
        # field for documenting why the reuse is happening
        description: Whenever a user is performing an action that affects another user
          -- typically in IAM scenarios -- the user affected by the action is captured at user.target
      - at: user
        as: effective
      - at: user
        as: changes
```

**Review by jonathan-buttner — DISMISSED** [review]
The proposal looks good to me. @rw-access any additional scenarios you can think of that would warrant additional fields?

**Comment by jonathan-buttner** [comment]
@rylnd not sure if you've seen anything from the SIEM that we'd need to include with these changes as well?

File: rfcs/text/0000-multiple-users.md (hunk 2 above)
Comment (webmat) [reply rfcs/text/0000-multiple-users.md:355]: Yes we will handle that a bit more loosely. We have work in progress (https://github.com/elastic/ecs/issues/932) to support a section of free form asciidoc documentation per field set (in addition to all of the attributes in the yaml).

Once we have the mechanism for this free form documentation section, the "user" field set documentation will lay out all expected uses with examples, based on what's in this RFC.

Comment (jonathan-buttner) [reply rfcs/text/0000-multiple-users.md:355]: Nice! Sounds good.

File: rfcs/text/0000-multiple-users.md [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    },
+    "target": {
+      "name": "bob",
+      "id": "1002"
+    },
+    "changes": {
+      "name": "bob.barker"
+    }
+  },
+  "related": { "user": ["alice", "root", "bob", "bob.barker"] }
```
Comment (rylnd) [inline rfcs/text/0000-multiple-users.md:279]: Apologies if this has already been addressed, but one concern that was raised by the detections team was the uniqueness of these names. If `related.user` is intended to be used programmatically, my (naive) sense is that `id` would be a better candidate here.

Comment (webmat) [reply rfcs/text/0000-multiple-users.md:279]: Yes, this has been raised in the past. We haven't addressed it directly anywhere yet, however. One potential solution that has been floated would be that `related.user` could perhaps contain all seen usernames + user IDs.

The field is only meant to pivot and find a user, no matter where in the event they appeared. It's not meant to have semantic meaning.

This RFC is not meant to address this issue, however. We will tackle that issue separately.

**Comment by webmat** [comment]
Ok I think this is ready for a final review for stage 2.

@janniten Thanks so much for these events, they're the ones I used for the Windows examples. Note that it's fine if events don't actually need all the field nestings at the same time. They're all in the schema to support events that do need them all. But in many cases, the activity is spread out across multiple events, as you noted. So only the needed fields actually need to be used :-) I've tried to demonstrate the various situations in the examples I've just added. 

@leehinman Thanks a lot for the Cloudtrail AssumeRole events. I modified one of the ARNs slightly, but I used your events in my cloud example :-) Can you confirm my mapping makes sense to you? Especially the last paragraph.

Final note for reviewers: I took the liberty to use a "session" `event.category` in some of the examples. It doesn't exist yet in the ECS allowed values, but has been discussed in the [session RFC](https://github.com/elastic/ecs/pull/935). 

@ebeahan  We have one additional merge task once this is approved:

- [ ] Set the merge date
- [ ] Assign the RFC ID
  - [ ] *Rename the directory with field changes accordingly* :-)

File: rfcs/text/0000-multiple-users.md [hunk 4]
```diff
@@ -282,6 +282,185 @@ the event now looks like:
 
 ## Source data
 
+Here are some concrete examples of events with multiple user and user roles.
+
+### Linux user creation
+
+Here's a typical set of log about a user creation on Linux.
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
Here's a typical set of logs about a user creation on Linux.
```

File: rfcs/text/0000-multiple-users.md [hunk 5]
```diff
@@ -282,6 +282,185 @@ the event now looks like:
...
+A solution that produces one event per log without coalescing would instead only
+represent users with the information available in the given log event.
+
+event 1:
+
+```JSON
+{
+  "event": {
+    "category": ["process"],
+    "event": "creation"
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": "creation"
```

File: rfcs/text/0000-multiple-users.md [hunk 6]
```diff
@@ -282,6 +282,185 @@ the event now looks like:
...
+}
+```
+
+event 2:
+
+```JSON
+{
+  "event": {
+    "category": ["session"],
+    "event": ["creation"],
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": ["creation"],
```

File: rfcs/text/0000-multiple-users.md [hunk 7]
```diff
@@ -282,6 +282,185 @@ the event now looks like:
...
+we capture the user at the root of the event. We do this despite the fact that
+they are not the one performing the action.
+
+event 5:
+
+```JSON
+{
+  "event": {
+    "category": ["session"],
+    "event": ["end"]
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": ["end"]
```

File: rfcs/text/0000-multiple-users.md [hunk 8]
```diff
@@ -460,10 +463,106 @@ event 5:
...
+
+Would translate to
+
+```JSON
+{
+  "event": {
+    "code": "4624",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "category": ["authentication"],
+    "event": ["start"],
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": ["start"],
```

File: rfcs/text/0000-multiple-users.md [hunk 9]
```diff
@@ -460,10 +463,106 @@ event 5:
...
+
+Would translate to
+
+```JSON
+{
+  "event": {
+    "code": "4781",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "category": ["iam"],
+    "event": ["user", "change"],
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": ["user", "change"],
```

File: rfcs/text/0000-multiple-users.md [hunk 10]
```diff
@@ -571,7 +571,69 @@ Would translate to
...
+
+And would translate to:
+
+```JSON
+{
+  "event": {
+    "action": "AssumeRole",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
Comment (ebeahan) [inline rfcs/text/0000-multiple-users.md]:
```suggestion
    "type": ["start"],
```

File: rfcs/0000-rfc-template.md [hunk 11]
```diff
@@ -83,7 +83,7 @@ The following are the people that consulted on the contents of this RFC.
 * TBD | author
 
 <!--
-Who will be or has consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.
+Who will be or has been consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.
```
Comment (ebeahan) [inline rfcs/0000-rfc-template.md:86]: 👍

File: rfcs/text/0007-multiple-users.md [hunk 12]
```diff
@@ -0,0 +1,769 @@
...
+## Source data
+
+Here are some concrete examples of events with multiple users and user roles.
+Note that the design of these fields is meant to allow all of their use at the
+same time, *when needed*. However if events don't contain all user roles because
+they're spread out across events, only the fields relevant to each event should be used.
+
+### Linux IAM and privilege escalation
+
+Here's a typical set of log about a user creation on Linux.
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:292]:
```suggestion
Here's a typical set of logs about a user creation on Linux.
```

File: rfcs/text/0007-multiple-users.md [hunk 13]
```diff
@@ -0,0 +1,769 @@
...
+represent users with the information available in the given log event.
+
+event 1:
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["process"],
+    "event": "creation"
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:373]:
```suggestion
    "type": ["creation"]
```

File: rfcs/text/0007-multiple-users.md [hunk 14]
```diff
@@ -0,0 +1,769 @@
...
+```
+
+event 2 (privilege escalation):
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["session"],
+    "event": ["creation"],
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:395]:
```suggestion
    "type": ["creation"],
```

File: rfcs/text/0007-multiple-users.md [hunk 15]
```diff
@@ -0,0 +1,769 @@
...
+they are not the one performing the action.
+
+event 5:
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["session"],
+    "event": ["end"]
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:462]:
```suggestion
    "type": ["end"]
```

File: rfcs/text/0007-multiple-users.md [hunk 16]
```diff
@@ -0,0 +1,769 @@
...
+Would translate to
+
+```JSON
+{
+  "event": {
+    "code": "4624",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:505]:
```suggestion
    "type": ["start"],
```

File: rfcs/text/0007-multiple-users.md [hunk 17]
```diff
@@ -0,0 +1,769 @@
...
+Would translate to
+
+```JSON
+{
+  "event": {
+    "code": "4781",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "kind": "event",
+    "category": ["iam"],
+    "event": ["user", "change"],
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:554]:
```suggestion
    "type": ["user", "change"],
```

File: rfcs/text/0007-multiple-users.md [hunk 18]
```diff
@@ -0,0 +1,769 @@
...
+
+And would translate to:
+
+```JSON
+{
+  "event": {
+    "action": "AssumeRole",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:618]:
```suggestion
    "type": ["start"],
```

**Review by ebeahan — APPROVED** [review]
LGTM 👍 

I only noted a handful of tiny changes.

File: rfcs/text/0007-multiple-users.md (hunk 13 above)
Comment (webmat) [reply rfcs/text/0007-multiple-users.md:373]: I will not drop a 🤦  on each of these, but the temptation is high 😂

Thanks for catching that.

File: rfcs/text/0007-multiple-users.md [hunk 19]
```diff
@@ -2,7 +2,7 @@
 <!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->
 
 - Stage: **2 (proposal)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
-- Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
+- Date: **2020-09-02** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
```
Comment (ebeahan) [inline rfcs/text/0007-multiple-users.md:5]: 🎉

**Review by ebeahan — APPROVED** [review]
LGTM :shipit:

**Comment by leehinman** [comment]
> Ok I think this is ready for a final review for stage 2.
> 
> @janniten Thanks so much for these events, they're the ones I used for the Windows examples. Note that it's fine if events don't actually need all the field nestings at the same time. They're all in the schema to support events that do need them all. But in many cases, the activity is spread out across multiple events, as you noted. So only the needed fields should be used in each such events :-) I've tried to demonstrate the various situations in the examples I've just added.
> 
> @leehinman Thanks a lot for the Cloudtrail AssumeRole events. I modified one of the ARNs slightly, but I used your events in my cloud example :-) Can you confirm my mapping makes sense to you? Especially the last paragraph.
> 
> Final note for reviewers: I took the liberty to use a "session" `event.category` in some of the examples. It doesn't exist yet in the ECS allowed values, but has been discussed in the [session RFC](https://github.com/elastic/ecs/pull/935).
> 
> @ebeahan We have one additional merge task once this is approved:
> 
> * [x]  Set the merge date
> * [x]  Assign the RFC ID
>   
>   * [x]  _Rename the directory with field changes accordingly_ :-)

LGTM.

**Pull request merged**